clippy_utils/src/check_proc_macro.rs

```diff
@@ -541,6 +541,102 @@ fn ident_search_pat(ident: Ident) -> (Pat, Pat) {
     (Pat::Sym(ident.name), Pat::Sym(ident.name))
 }
 
+fn pat_search_pat(tcx: TyCtxt<'_>, pat: &rustc_hir::Pat<'_>) -> (Pat, Pat) {
+    match pat.kind {
+        rustc_hir::PatKind::Missing | rustc_hir::PatKind::Err(_) => (Pat::Str(""), Pat::Str("")),
+        rustc_hir::PatKind::Wild => (Pat::Sym(kw::Underscore), Pat::Sym(kw::Underscore)),
+        rustc_hir::PatKind::Binding(binding_mode, _, ident, Some(end_pat)) => {
+            let start = match binding_mode.0 {
+                rustc_hir::ByRef::Yes(rustc_hir::Mutability::Not) => Pat::Str("ref"),
+                rustc_hir::ByRef::Yes(rustc_hir::Mutability::Mut) => Pat::Str("ref mut"),
+                rustc_hir::ByRef::No => {
+                    let (s, _) = ident_search_pat(ident);
+                    s
+                },
+            };
```

Comment:
This can be something like

```rust
if binding_mode == BindingMode::NONE {
    ident_search_pat(ident).0
} else {
    Pat::Str(binding_mode.prefix_str())
}
```